nucleus/dataset_item.py

```diff
@@ -2,36 +2,135 @@
 import json
 import os.path
 from dataclasses import dataclass
-from typing import Optional, Sequence
-from urllib.parse import urlparse
+from typing import Optional, Sequence, Dict, Any
+from enum import Enum
 
+from .annotation import is_local_path, Point3D
 from .constants import (
     DATASET_ITEM_ID_KEY,
     IMAGE_URL_KEY,
     METADATA_KEY,
     ORIGINAL_IMAGE_URL_KEY,
     REFERENCE_ID_KEY,
+    TYPE_KEY,
+    URL_KEY,
+    CAMERA_PARAMS_KEY,
+    POINTCLOUD_URL_KEY,
+    X_KEY,
+    Y_KEY,
+    Z_KEY,
+    W_KEY,
+    POSITION_KEY,
+    HEADING_KEY,
+    FX_KEY,
+    FY_KEY,
+    CX_KEY,
+    CY_KEY,
 )
 
 
 @dataclass
-class DatasetItem:
+class Quaternion:
+    x: float
+    y: float
+    z: float
+    w: float
 
-    image_location: str
+    @classmethod
+    def from_json(cls, payload: Dict[str, float]):
+        return cls(
+            payload[X_KEY], payload[Y_KEY], payload[Z_KEY], payload[W_KEY]
+        )
+
+    def to_payload(self) -> dict:
+        return {
+            X_KEY: self.x,
+            Y_KEY: self.y,
+            Z_KEY: self.z,
+            W_KEY: self.w,
+        }
+
+
+@dataclass
+class CameraParams:
+    position: Point3D
+    heading: Quaternion
+    fx: float
+    fy: float
+    cx: float
+    cy: float
+
+    @classmethod
+    def from_json(cls, payload: Dict[str, Any]):
+        return cls(
+            Point3D.from_json(payload[POSITION_KEY]),
+            Quaternion.from_json(payload[HEADING_KEY]),
+            payload[FX_KEY],
+            payload[FY_KEY],
+            payload[CX_KEY],
+            payload[CY_KEY],
+        )
+
+    def to_payload(self) -> dict:
+        return {
+            POSITION_KEY: self.position.to_payload(),
+            HEADING_KEY: self.heading.to_payload(),
+            FX_KEY: self.fx,
+            FY_KEY: self.fy,
+            CX_KEY: self.cx,
+            CY_KEY: self.cy,
+        }
+
+
+class DatasetItemType(Enum):
+    IMAGE = "image"
+    POINTCLOUD = "pointcloud"
+
+
+@dataclass  # pylint: disable=R0902
+class DatasetItem:  # pylint: disable=R0902
+    image_location: Optional[str] = None
     reference_id: Optional[str] = None
     item_id: Optional[str] = None
     metadata: Optional[dict] = None
+    pointcloud_location: Optional[str] = None
 
     def __post_init__(self):
         self.local = is_local_path(self.image_location)
+        assert bool(self.image_location) != bool(
```

Comment:
Need to check that exactly one of 

image_location
pointcloud_location

is passed

Reply:
I believe that this code does this - ` bool(self.image_location) != bool(self.pointcloud_location)` essentially performs an XOR so only 1 has been set